nixos/modules/services/mail/dovecot.nix

```
@@ -272,22 +210,195 @@ let
+      (mkRemovedOptionModule [
+        "services"
+        "dovecot2"
+        "sslServerKey"
+      ] "Use `settings.ssl_key` for Dovecot 2.3, `settings.ssl_server_key_file` for 2.4.")
+      (mkRemovedOptionModule [
+        "services"
+        "dovecot2"
+        "sslCACert"
+      ] "Use `settings.ssl_ca` for Dovecot 2.3, `settings.ssl_server_ca_file` for 2.4.")
+      (mkRemovedOptionModule [
+        "services"
+        "dovecot2"
+        "mailLocation"
+      ] "Use `settings.mail_location` for Dovecot 2.3, `settings.mail_home` for 2.4.")
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "sieveScripts" ]
+        [ "services" "dovecot2" "sieve" "scripts" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "mailUser" ]
+        [ "services" "dovecot2" "settings" "mail_uid" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "mailGroup" ]
+        [ "services" "dovecot2" "settings" "mail_gid" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "protocols" ]
+        [ "services" "dovecot2" "settings" "protocols" ]
+      )
+    ]
+    ++ (
+      let
+        basePath = [
+          "services"
+          "dovecot2"
+        ];
+        mkRemovedOptions =
+          list:
+          map (
+            name: mkRemovedOptionModule (basePath ++ name) "Please use services.dovecot2.settings instead."
+          ) list;
+      in
+      mkRemovedOptions [
+        [ "pluginSettings" ]
+        [ "enableQuota" ]
+        [ "quotaPort" ]
+        [ "quotaGlobalPerUser" ]
+      ]
+    );
 
   options.services.dovecot2 = {
     enable = mkEnableOption "the dovecot 2.x POP3/IMAP server";
 
+    package = mkPackageOption pkgs "dovecot" { } // {
+      default =
+        if versionAtLeast config.system.stateVersion "25.11" then pkgs.dovecot else pkgs.dovecot_2_3;
+    };
+
+    settings = mkOption {
+      default = { };
+      type =
+        let
+          inherit (lib.types)
+            attrsOf
+            bool
+            int
+            nonEmptyListOf
+            nonEmptyStr
+            nullOr
+            oneOf
+            str
+            submodule
+            ;
+          inherit (lib.lists) last dropEnd;
+
+          section = submodule (
+            { name, options, ... }:
+            let
+              # if the current name is a list (starting with '[definition ') -> get
+              # name' from _module.args.loc & use it for {type,name}Default
+              name' =
+                if (builtins.match ''\[definition .*'' name) == null then
+                  name
+                else
+                  last (dropEnd 3 options._module.args.loc);
+
+              # split name' on the first space
+              splits = builtins.match "([^ ]+) +(.+)" name';
+              typeDefault = if splits == null then name' else builtins.elemAt splits 0;
+              nameDefault = if splits == null then null else builtins.elemAt splits 1;
```

As per https://github.com/NixOS/nixpkgs/pull/410912#discussion_r2107806459, should we drop the ` +` ?

```suggestion
              # split name' on the first space
              splits = builtins.match "([^ ]+) (.+)" name';
              typeDefault = if splits == null then name' else builtins.elemAt splits 0;
              nameDefault = if splits == null then null else builtins.elemAt splits 1;
```